### PR TITLE
Graceful invalidation of Python Node/Value/Block when C++ object is deleted

### LINKED ISSUE #49969: Improve JIT Python interface safety
## 🚀 Feature
I propose to improve the Python JIT interface safety by wrapping the "custom ownership" C++ classes like Node, Value, and Block in smart pointers supporting invalidation.

## Motivation

Currently the JIT Python bindings for Node, Block, Value and potentially others (where the ownership model is "owned by graph, handed out as raw pointers, but may be destroyed at will" work with essentially raw pointers, making it too easy to access freed memory, which is bad (and might lead to crashes).

## Pitch

(At least) for the Python bindings we should wrap the objects using a custom shared smart pointer which supports invalidation that is done when destroying the C++ object wrapped by the Python object.

This would involve (e.g. for a Node):
- create a wrapper (struct with a (raw) pointer to a Node) when creating a node, hold on to that with a shared_ptr,
- set the wrapper's pointer to the node to `nullptr` on destruction,
- cause PyBind11 to check and dereference the wrapper at each invocation. This can be done using a custom caster (see https://github.com/pybind/pybind11/issues/2751).

This won't give us thread safety, but at least we don't crash from Python by using Python Node objects pointing to invalid memory.

## Alternatives

- Keep stuff unsafe.
- Move to some smart-pointer thing for C++, too, and use some invalidation scheme there.

## Additional context

This would cause the creation of an additional struct and a smart pointer to it in each instantiation of a Node (etc.) and clearing the pointer on destruction, which would be the main perf impact. All other changes only affect the Python bindings, which I would think that is less critical.

I do have experimental code doing this and am looking to be able to build a subgraph-creating pass (say, similar to `create_autodiff_subgraphs` from Python).

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by t-vi**
Previously we might have gotten segfaults and all, now it raises an exception.
Thread safety hasn't been an objective.

I have a followup to expand the Python interface for the API.

Fixes #49969.

@wanchaol

### CODE REVIEW

**Comment by t-vi** [comment]
@suo @eellison from the tagging of the issue.

**Comment by t-vi** [comment]
The error looks rather serious, I'll have to check what that is about.

**Comment by t-vi** [comment]
Seems like the error was transient after all (and I couldn't reproduce locally, too). Maybe ONNX export does something with a race condition?

**Comment by t-vi** [comment]
It turns out that PyBind seems to still cache entries based the raw pointer we're wrapping. This is bad becasue the next object from the new reference will then be "accessed" through the old wrapper.

**Comment by t-vi** [comment]
I added invalidating the cache. Now it runs like a charm.

**Comment by t-vi** [comment]
Any chance someone could give an initial direction for the patch?

**Comment by suo** [comment]
I'll find someone to take a look (or look myself if I get a chance in the next few days). Thanks for the ping!

**Comment by t-vi** [comment]
Thank you!

File: torch/csrc/jit/python/pybind.h [hunk 1]
```diff
@@ -17,6 +17,90 @@
...
+
+// This is a variant of shared_ptr that "sees through" a wrapper.
+// We use it to convert Value, Node, Block and node to "wrapped" Python
+// values. When we destruct the C++ object, the wrapper's pointer will
+// be set to 0 and any future dereferencing will throw. We need this
+// because the Python objects may hang around after the C++ object
+// has already been destroyed.
+// This also needs the magic type_caster below, which is from the
+// workaround offered in https://github.com/pybind/pybind11/issues/2751
+template <typename T>
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/pybind.h:31]: Should we add some static checks here to ensure T is one of the IR structs?

**Review by gmagogsfm — COMMENTED** [review]
Looks good. I imported it to FB internal for testing.

One thing that I am curious though, is this problem (memory invalidated in C++, leaving empty reference for PyBind) unique to TorchScript?

File: torch/csrc/jit/python/pybind.h (hunk 1 above)
Comment (t-vi) [reply torch/csrc/jit/python/pybind.h:31]: Sounds like a good idea. Thank you!

**Comment by t-vi** [comment]
> One thing that I am curious though, is this problem (memory invalidated in C++, leaving empty reference for PyBind) unique to TorchScript?

So the stock answer here is "don't do that, keep alive the C++ object as long as the Python object lives" (see also the discussion on the PyBind issue). We do this e.g. with StrongFunctionPtr for ScriptFunctions. And this makes a lot of sense most of the time.
In our situation:
- we have a rather complex ownership model, where Nodes etc. become "invalid" when transformations destroy them,
- we want to keep the raw pointers in C++ and not introduce valid-checks in every method,
- I didn't want to introduce a proxy object with methods that does that just for Python.

If these weren't the constraints, we would probably make the Node* shared pointers and check if they're valid whenever we use them. This would make it again trivial to wrap in Python.
Or having making the a NodeWrapper instead of `Wrapper<Node>` and passing through all functions we want to expose in Python unwrapping the inputs and wrapping the outputs would make the Python bit trivial again, too, without touching C++. One could probably make this less manual by some C++ argument forwarding regime, but then we're basically reinventing half of PyBind.

File: torch/csrc/jit/python/pybind.h (hunk 1 above)
Comment (t-vi) [reply torch/csrc/jit/python/pybind.h:31]: I added a static assert. Again, thank you for the suggestion!

**Comment by t-vi** [comment]
It seems the FB build & test is somehow stuck.

**Comment by gmagogsfm** [comment]
> It seems the FB build & test is somehow stuck.

Yeah, there seems to be a transient failure. Just kicked off the testing again.

**Comment by t-vi** [comment]
Thank you @gmagogsfm !

**Comment by t-vi** [comment]
Now it's failing after 40 hours...

**Comment by gmagogsfm** [comment]
> Now it's failing after 40 hours...

Just checked the errors, they don't look anything related to this diff. The errors are just failing to read some temporary files, my guess is it is infra issue. Let me retry one more time.

**Comment by t-vi** [comment]
Given that it failed again, should I rebase?

**Comment by gmagogsfm** [comment]
> Given that it failed again, should I rebase?

Don't worry, I will take care of rebasing.

After a few retries, I think this may be more than just infra failures. Trying out some manual test runs internally to see what's going on. Will update here.

**Comment by gmagogsfm** [comment]
> > Given that it failed again, should I rebase?
> 
> Don't worry, I will take care of rebasing.
> 
> After a few retries, I think this may be more than just infra failures. Trying out some manual test runs internally to see what's going on. Will update here.

@t-vi 

Just did some investigation, the FB internal tests failure was a mix of both infra failure and some real segfaults. To reproduce same error, just sync `pybind11` to v2.2.1 and build PyTorch with it. Then pretty much any tests would fail with a segfault.

**Comment by t-vi** [comment]
But so is the goal to make it work with old PyBind (2.2.1 is from 2017) or is that just a way of reproducing something that is also present in newer?

**Comment by gmagogsfm** [comment]
> But so is the goal to make it work with old PyBind (2.2.1 is from 2017) or is that just a way of reproducing something that is also present in newer?
> 
> One thing I noticed is that PyTorch appears to link against system pybind if it is installed in /usr/include rather than the thirdpart/pyind version.

The issue is not present in newer. If we could make this PR work well with old 2.2.1, that would be ideal. Upgrading internal pybind to newer versions will take a long time and it blocks this PR. I will also research to see if there is an easier way of doing it.

**Comment by t-vi** [comment]
OK, thank you for putting this into perspective. I can reproduce a segfault and will look into it!

**Comment by t-vi** [comment]
So the fix the old version doesn't have is this: https://github.com/pybind/pybind11/pull/1435
Apparently there is a general problem with custom smart pointers before this.

**Comment by t-vi** [comment]
For me locally, just disabling the `&` overload for old pybind seems to do the trick. This seems plausible as not too many things in old PyBind would use it if it leads to segfaults.

**Comment by gmagogsfm** [comment]
> For me locally, just disabling the `&` overload for old pybind seems to do the trick. This seems plausible as not too many things in old PyBind would use it if it leads to segfaults.

Thanks! Internal tests look a lot better now, trying to merge.

**Comment by t-vi** [comment]
Thank you, @gmagogsfm !

**Pull request closed without merging**